File: source_code_files/C#/dp_knapsack_top-down_recursion.cs
// dp_knapsack_top-down_recursion.cs
//
// Robert Sackmann, 2023-09-14
//


using System.Diagnostics;  // Stopwatch


namespace dp_knapsack_top_down;


class knapsack_recursive
{
    int max_wt;  // current weight capacities
    int[]? picks;  // Consider declaring the field as nullable.

    public static int result = -1;

    static long time_counter;
    static long int_milli_inter;
    Stopwatch? timeout_timer;  // Consider declaring the field as nullable.
    static bool CLEANUP_FLAG;  // for timeout calculation

    // const long TIME_LIMIT = 120000;  // = 2 min
    const long TIME_LIMIT = 1200000;  // = 20 min
    const long TIME_CHECK_COUNTER = 30000;  // tick counter in recursive function for timeout check


    // recursive function:
    int solveKnapsack (
        int[] weight,
        int[] val,
        int[] picks,
        int items,
        int capacity
        ) {

        int include, exclude;

        if (time_counter > TIME_CHECK_COUNTER) {
            if (timeout_timer != null) {
                int_milli_inter = timeout_timer.ElapsedMilliseconds;  // Dereference of a possibly null reference.

                if (int_milli_inter > TIME_LIMIT) {
                   if (CLEANUP_FLAG == false) {
                       CLEANUP_FLAG = true;
                       Console.WriteLine("\n  >>> Dynamic Programming top-down: stopping execution due to exceeded time limit of 20 minutes. Cleaning up...\n");
                   }

                   return Int32.MinValue;
                }
                else {
                    time_counter = 0;
                }
            }
        }

        time_counter++;


        if (capacity < 0) return Int32.MinValue;
        if (items < 0 || capacity == 0) return 0;

        include = val[items] + solveKnapsack(weight, val, picks, items-1, capacity-weight[items]);

        exclude = solveKnapsack(weight, val, picks, items-1, capacity);

        // building up the (huge) picks table (2dim array):
        if (include > exclude) {
            picks[items*max_wt + capacity] = 1;
        }
        else {
            picks[items*max_wt + capacity] = -1;
        }

        return Math.Max(include, exclude);
    }


    int printPicks(
        int[] weight,
        int[] val,
        int[] picks,
        int N,
        int capacity) {

        int act_total_weight = 0;
        int act_total_value = 0;
        int size = capacity;
        int item = N-1;  // item id = 0...N-1

        Console.WriteLine("\n  Optimal weights:");

        while (item >= 0 && act_total_weight < capacity) {
            if (picks[item*capacity + size] == 1) {
                Console.WriteLine("    id = " + item + ",  weight = " + weight[item] + ",  value = " + val[item]);
                act_total_weight += weight[item];
                act_total_value += val[item];
                size -= weight[item];
            }
            item -= 1;
        }

        Console.WriteLine("  => actual weight of optimal knapsack = " + act_total_weight);
        Console.WriteLine("  => actual value of optimal knapsack = " + act_total_value);

        return 0;
    }



    public knapsack_recursive (int[] w, int[] v, int T)
    {
        int n = w.Length;

        max_wt = T;

        ulong picks_alloc = (Convert.ToUInt64(n))*(Convert.ToUInt64(T+1));  // 01_WEIGHTS4.in: (4)*(7+1) = 32

        Console.WriteLine("\n  DP top-down with picks table on: number of elements of picks table = " + picks_alloc);

        // ulong picks_alloc_bytes = picks_alloc * 4;  // speculative!!
        // Console.WriteLine("                                   number of bytes of picks table =    " + picks_alloc_bytes);

        /*
        <gcAllowVeryLargeObjects> element (2023-09-01)
        https://learn.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/runtime/gcallowverylargeobjects-element?redirectedfrom=MSDN
          The maximum size in any single dimension is 2,147,483,591 (0x7FFFFFC7) for byte arrays and
          arrays of single-byte structures, and 2,146,435,071 (0X7FEFFFFF) for arrays containing other types.
        If not considered --> Unhandled exception. System.OverflowException: Array dimensions exceeded supported range; Aborted (core dumped)

        1 int value takes 4 bytes: https://learn.microsoft.com/de-de/dotnet/api/system.int32?view=net-7.0

        free available memory: do not calculate it here with C#: implicitly assume that there would be always something like 4GB or so for the memory: System.Int32.MaxValue is the most probable limit:

        Test case: 05_WEIGHTS_TODD_18.in:
            max capacity of knapsack = 79691769
            DP top-down with picks table on: number of elements of picks table = 1434451860
                                                number of bytes of picks table =    5737807440
                  (maximum size of arrays of non-single-byte structures [bytes] =    2146435071) --> speculative!!
            DP top-down: start recursion...
            end of recursion.
        */

        try {
            picks = new int[picks_alloc];

            // start timeout timer:
            timeout_timer = Stopwatch.StartNew();

            time_counter = 0;
            CLEANUP_FLAG = false;

            Console.WriteLine("\n  DP top-down: start recursion...");
            result = solveKnapsack(w, v, picks, n-1, max_wt); // n-1, not n: stay within the bounds of the w and v arrays!!
            Console.WriteLine("  end of recursion.");

            timeout_timer.Stop();

            if (CLEANUP_FLAG == false) {
                int result_pr = printPicks(w, v, picks, n, max_wt);
            }
        }
        catch (Exception e) {
            Console.WriteLine("\n  Can't allocate memory for an int array with elements for a picks table of needed size: " + e.Message);
            result = -1;
        }

        if (CLEANUP_FLAG == true) {
            result = -1;
        }
    }
}



class knapsack_recursive_no_picks
{
    int max_wt;  // current weight capacities

    public static int result = -1;

    static long time_counter;
    static long int_milli_inter;
    Stopwatch? timeout_timer;  // Consider declaring the field as nullable.
    static bool CLEANUP_FLAG;  // for timeout calculation

    // const long TIME_LIMIT = 120000;  // = 2 min
    const long TIME_LIMIT = 1200000;  // = 20 min
    const long TIME_CHECK_COUNTER = 30000;  // tick counter in recursive function for timeout check


    // recursive function:
    int solveKnapsack (
        int[] weight,
        int[] val,
        int items,
        int capacity
        ) {

        int include, exclude;

        if (time_counter > TIME_CHECK_COUNTER) {
            if (timeout_timer != null) {
                int_milli_inter = timeout_timer.ElapsedMilliseconds;  // Dereference of a possibly null reference.

                if (int_milli_inter > TIME_LIMIT) {
                   if (CLEANUP_FLAG == false) {
                       CLEANUP_FLAG = true;
                       Console.WriteLine("\n  >>> Dynamic Programming top-down: stopping execution due to exceeded time limit of 20 minutes. Cleaning up...\n");
                   }

                   return Int32.MinValue;
                }
                else {
                    time_counter = 0;
                }
            }
        }

        time_counter++;


        if (capacity < 0) return Int32.MinValue;
        if (items < 0 || capacity == 0) return 0;

        include = val[items] + solveKnapsack(weight, val, items-1, capacity-weight[items]);

        exclude = solveKnapsack(weight, val, items-1, capacity);

        return Math.Max(include, exclude);
    }


    public knapsack_recursive_no_picks (int[] w, int[] v, int T)
    {
        int n = w.Length;

        max_wt = T;

        timeout_timer = Stopwatch.StartNew();

        Console.WriteLine("\n  DP top-down: start recursion...");
        result = solveKnapsack(w, v, n-1, max_wt);  // n-1, not n: stay within the bounds of the w and v arrays!!
        Console.WriteLine("  end of recursion.");

        if (CLEANUP_FLAG == true) {
            result = -1;
        }
    }
}

File: source_code_files/C++/dp_knapsack_top-down_recursion.cpp
// dp_knapsack_top-down_recursion.cpp
//
// 2023-09-16
//


#include <iostream>
#include <vector>
#include <chrono>
#include <limits.h>  // INT_MIN


using namespace std;

using std::chrono::high_resolution_clock;
using std::chrono::duration;
using std::chrono::duration_cast;


#include <unistd.h>  // sysconf()
#include <sys/sysinfo.h>  // get_avphys_pages()


const long long MARGIN_OF_SAFETY_BYTES = 500000000;  // ~500 megabytes; only a very wild guess after some experiments

const long long TIME_LIMIT_MIN = 60000;  // = 1 min minimum timeout waiting time


class knapsack_recursive{
    vector<int> w;  // weights
    vector<int> v;  // values (prices)

    int T, max_wt;  // current weight capacities
    
    long long timeout_input;  // milliseconds

    vector<int> picks;  // 1dim array as vector (dynamic array) --> can grow dynamically

    long long int_milli_start, int_milli_inter, time_counter;
    int CLEANUP_FLAG;  // for timeout calculation

    // const double TIME_LIMIT = 300000;  // = 5 min
    // const long long TIME_LIMIT = 1200000;  // = 20 min

    const long long TIME_CHECK_COUNTER = 30000;  // tick counter in recursive function for timeout check


    // recursive function:
    int solveKnapsack(
        vector<int>& weight,
        vector<int>& value,
        vector<int>& picks,
        int items,
        int capacity) {

        int include, exclude;

        if (time_counter > TIME_CHECK_COUNTER) {
             auto t_i = high_resolution_clock::now();
             int_milli_inter = duration_cast<std::chrono::milliseconds>(t_i.time_since_epoch()).count();

             double duration = int_milli_inter - int_milli_start;

             if (duration > timeout_input)
             {
                if (CLEANUP_FLAG == 0) {
                    CLEANUP_FLAG = 1;
                    // cout << "\n  >>> Dynamic Programming top-down: stopping execution due to exceeded time limit of 20 minutes. Cleaning up...\n" << endl;
                    cout << "\n  >>> Dynamic Programming top-down: stopping execution due to exceeded time limit of " << timeout_input / TIME_LIMIT_MIN << "minutes. Cleaning up...\n" << endl;
                }

                return INT_MIN;
             }
             else time_counter = 0;
         }

         time_counter++;

         if (capacity < 0) return INT_MIN;
         if (items < 0 || capacity == 0) return 0;

         include = value[items] + solveKnapsack(weight, value, picks, items-1, capacity-weight[items]);

         exclude = solveKnapsack(weight, value, picks, items-1, capacity);

         // building up the (huge) picks table (2dim array):
         if (include > exclude) {
             picks[items*max_wt + capacity] = 1;
         }
         else {
             picks[items*max_wt + capacity] = -1;
         }

        return max(include, exclude);
    }


    // print the optimal weights and the actual total weight of the optimal knapsack:
    int printPicks(
        vector<int>& weight,
        vector<int>& value,
        vector<int>& picks,
        int N,
        int capacity) {

        int act_total_weight = 0;
        int act_total_value = 0;
        int size = capacity;
        int item = N-1;  // item id = 0...N-1

        cout << "\n  Optimal weights as picked by the Dynamic Programming top-down (recursive) algo:" << endl;

        while (item >= 0 && act_total_weight < capacity) {
            if (picks[item*capacity + size] == 1) {
                cout << "    id = " << item << ",  weight = " << weight[item] << ",  value = "<< value[item] << endl;
                act_total_weight += weight[item];
                act_total_value += value[item];
                size -= weight[item];
            }
            item -= 1;
        }

        cout << "  => actual weight of optimal knapsack = " << act_total_weight << endl;
        cout << "  => actual value of optimal knapsack = " << act_total_value << endl;

        return 0;
    }


public:
    int result, result_pr = 0;

    knapsack_recursive(vector<int>& w, vector<int>& v, int T, long long timeout_input) : w(w), v(v), T(T), timeout_input(timeout_input) {
        int n = w.size();

        max_wt = T;

        // number of elements of picks table:
        // 2023-09-10: only n * (T+1) elements needed (not (n+1) * (T+1))
        long long picks_alloc = (static_cast<long long>(n)) * (static_cast<long long>(T)+1);


        cout << "\n  DP top-down with picks table on: number of elements of picks table = " << picks_alloc << endl;

        int size_int = sizeof(int);  // vector<int> picks;
        long long picks_alloc_bytes = picks_alloc * static_cast<long long>(size_int);
        cout << "                                   number of bytes of picks table =    " << picks_alloc_bytes << endl;

        // Test case: ./06_WEIGHTS_TODD_19.in: picks_alloc = (19 + 1) * (167772154 + 1) = 3355443100 (elements of type int)
        //   => 3355443100 * 4 bytes = 13421772400 bytes = ~13.4 gigabytes <-- picks_alloc_bytes
        // https://convertlive.com/u/convert/bytes/to/gigabytes

        // Test case: ./06_WEIGHTS_TODD_18.in: picks_alloc_bytes = 6056574520 = ~6.06 gigabytes

        // If an exception is thrown for any reason, these functions have no effect (strong exception safety guarantee). Although not explicitly specified, std::length_error is thrown if the capacity required by the new vector would exceed max_size().
        // from: https://en.cppreference.com/w/cpp/container/vector/resize
        //
        // https://stackoverflow.com/questions/23678678/c-vector-catch-resize-memory-leak
        // this doesn't work with 06_WEIGHTS_TODD_19.in: the recursion is starting, only to see after some while: Segmentation fault (core dumped):
        //   try
        //   {
        //       picks.resize(picks_alloc,0);
        //   }
        //   catch (const bad_alloc& ba)
        //   {
        //       cerr << "\nMemory Exhaustion! Leaving this test case..." << endl;
        //   }

        long long avl_pages = get_avphys_pages();  // available physical page count
        long long page_size = sysconf(_SC_PAGE_SIZE);
        long long avl_page_size_bytes = avl_pages * page_size;
        cout << "                                   avl_page_size_bytes =               " << avl_page_size_bytes << " (available free memory in bytes)" << endl;

        long long net_free_memory_guess = avl_page_size_bytes - MARGIN_OF_SAFETY_BYTES;
        cout << "  available free memory after applying a margin of safety =            " << net_free_memory_guess << endl;


        if (picks_alloc_bytes > net_free_memory_guess) {
            cerr << "\n  recursion is not starting here: it may run out of free available memory including some safety margin!" << endl;

            result = -1;
        }
        else {
            picks.resize(picks_alloc,0);

            // start timeout timer:
            auto t_start = high_resolution_clock::now();
            int_milli_start = duration_cast<std::chrono::milliseconds>(t_start.time_since_epoch()).count();

            time_counter = 0;
            CLEANUP_FLAG = 0;

            cout << "\n  DP top-down: start recursion..." << endl;
            result = solveKnapsack(w, v, picks, n-1, max_wt);  // n-1, not n: stay within the bounds of the w, v arrays!!
            cout << "  end of recursion." << endl;

            if (CLEANUP_FLAG == 1) {
                result = -1;
            }
            else {
                result_pr = printPicks(w, v, picks, n, max_wt);
            }
        }
    }
};



class knapsack_recursive_no_picks{
    vector<int> w;  // weights
    vector<int> v;  // values (prices)

    int T, max_wt;  // current weight capacities
    
    long long timeout_input;  // milliseconds

    long long int_milli_start, int_milli_inter, time_counter;
    int CLEANUP_FLAG;  // for timeout calculation

    // const double TIME_LIMIT = 300000;  // = 5 min
    // const long long TIME_LIMIT = 1200000;  // = 20 min

    const long long TIME_CHECK_COUNTER = 30000;  // millisecond counter in recursive function for timeout check


    // recursive function:
    int solveKnapsack(
        vector<int>& weight,
        vector<int>& value,
        int items,
        int capacity) {

        int include, exclude;

        if (time_counter > TIME_CHECK_COUNTER) {
            auto t_i = high_resolution_clock::now();
            int_milli_inter = duration_cast<std::chrono::milliseconds>(t_i.time_since_epoch()).count();

            double duration = int_milli_inter - int_milli_start;

            if (duration > timeout_input)
            {
                if (CLEANUP_FLAG == 0) {
                    CLEANUP_FLAG = 1;
                    // cout << "\n  >>> Dynamic Programming top-down: stopping execution due to exceeded time limit of 20 minutes. Cleaning up...\n" << endl;
                    cout << "\n  >>> Dynamic Programming top-down: stopping execution due to exceeded time limit of " << timeout_input / TIME_LIMIT_MIN << "minutes. Cleaning up...\n" << endl;
                }

                return INT_MIN;
            }
            else time_counter = 0;
        }

        time_counter++;

        if (capacity < 0) return INT_MIN;
        if (items < 0 || capacity == 0) return 0;

        include = value[items] + solveKnapsack(weight, value, items-1, capacity-weight[items]);

        exclude = solveKnapsack(weight, value, items-1, capacity);

        return max(include, exclude);
    }


public:
    int result, result_pr = 0;

    knapsack_recursive_no_picks(vector<int>& w, vector<int>& v, int T, long long timeout_input) : w(w), v(v), T(T), timeout_input(timeout_input) {
        int n = w.size();

        max_wt = T;

        auto t_start = high_resolution_clock::now();
        int_milli_start = duration_cast<std::chrono::milliseconds>(t_start.time_since_epoch()).count();

        time_counter = 0;
        CLEANUP_FLAG = 0;

        cout << "\n  DP top-down: start recursion with picks table off..." << endl;
        result = solveKnapsack(w, v, n-1, max_wt);  // n-1, not n: stay within the bounds of the w, v arrays!!
        cout << "  end of recursion." << endl;

        if (CLEANUP_FLAG == 1) result = -1;
    }
};
